Sprint12

irinadeeva opened the pull request:
Created a network layer for the screen with a ribbon of photos.
Implemented the functionality of likes.
Added full-screen viewing of photos.
Added the ability to log out of the account.

Commit:
Сетевой слой для экрана с лентой фотографий

Commit:
image list service integration

Commit:
implemented a UI layer for a screen with a feed of photos

Commit:
did post request for photos/:id/like

Commit:
implementation UI for lIke status

Commit:
Scrollview: полноразмерный просмотр картинки

Commit:
update alert, logout
adapt alert 1 buttons and more
and clean token

Commit:
stub for Single Image

Commit:
clean

helecta commented on ImageFeed/Helper/String+Extensions.swift:
```diff
@@ -0,0 +1,18 @@
+}()
+
+extension String {
+    var changeToDate: Date? { dateTimeDefaultFormatter.date(from: self) }
```
Не очень удачное название свойства, похоже больше на метод.
Лучше переименовать в `formattedDate`

helecta commented on ImageFeed/ImageList/ImageListCell.swift:
```diff
@@ -6,10 +6,33 @@
+    func setIsLiked(isLiked: Bool) {
+        let buttonImage = isLiked == true ? UIImage(named: "Active") : UIImage(named: "No Active")
+
+        self.favouriteButton.setImage(buttonImage, for: .normal)
```
Тут `self` необязательно использовать, поэтому лучше его убрать.
Подробнее можешь почитать в [код стайле](https://github.com/tutu-ru/swift-style-guide?tab=readme-ov-file#Комментарии)

helecta commented on ImageFeed/ImageList/ImagesListViewController.swift:
```diff
@@ -94,4 +184,24 @@ extension ImagesListViewController : UITableViewDelegate {
+        imagesListService.changeLike(photoId: photo.id, isLike: !photo.isLiked) { result in
+            switch result {
+            case .success:
+                self.photos = self.imagesListService.photos
```
В замыкание для избежания retain cycl'ов, необходимо передавать слабую ссылку на `self`

helecta commented on ImageFeed/ImageList/ImagesListViewController.swift:
```diff
@@ -21,41 +25,87 @@ final class ImagesListViewController: UIViewController {
+
+    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        if indexPath.row + 1 == photos.count {
+            imagesListService.fetchPhotosNextPage()
```
Можно убрать притормаживание таблицы, если убрать загрузку из главного потока:
```
            DispatchQueue.global().async { [weak self] in
                self?.imagesListService.fetchPhotosNextPage() { result in }
            }
```
Не забудь убрать `assert(Thread.isMainThread)` из `fetchPhotosNextPage`

Обычно такого рода получение данных делают в фоне

helecta commented on ImageFeed/Models/LikeResultResponse.swift:
```diff
@@ -0,0 +1,26 @@
+    let id: String
+    let isLiked: Bool
+
+    private enum CodingKeys: String, CodingKey {
```
Еще можно таким образом настроить JSONDecoder: https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable  
Тогда CodingKeys будет не нужно

helecta commented on ImageFeed/ImageList/ImagesListViewController.swift:
```diff
@@ -21,41 +25,87 @@ final class ImagesListViewController: UIViewController {
+        }
 
-        let buttonImage = (indexPath.row % 2) == 0 ? UIImage(named: "Active") : UIImage(named: "No Active")
+        cell.dataLabel.text =  dateFormatter.string(from: photo.createdAt ?? Date())
```
Тут не стоит использовать по умолчанию текущую дату, т.к это будет уже отображение неправильных данных, по заданию тут должен быть пустой лейбл:
> либо в структуре Photo поле `createdAt` опционально, в ячейке лейбл пустой

helecta commented on ImageFeed/ImageList/ImagesListViewController.swift:
```diff
@@ -6,10 +6,14 @@
+    private var alertPresenter: AlertProtocol?
     private lazy var dateFormatter: DateFormatter = {
         let formatter = DateFormatter()
         formatter.dateStyle = .long
```
Если так настроить дата форматтер:
```
private lazy var dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd MMMM yyyy"
        formatter.locale = Locale(identifier: "ru_RU")
        return formatter
    }()
```
То будет выглядеть как в макете, без  `г.`

helecta commented on ImageFeed/SplashViewController/SplashViewController.swift:
Сейчас у тебя такой поток авторизации:
1. Приложение запускается, показан сплеш экран
2. Если пользователь не авторизован, показывается экран авторизации
3. Пользователь авторизовываются во вебвью
4. Вебвью закрывается и показывается сплеш скрин с лоудингом
5. Успевает показаться еще раз экран авторизации перед тем как будет показана лента

Это происходит потому что в SplashScreenViewController, когда в методе `func authViewController(_ vc: AuthViewController, didAuthenticateWithCode code: String)` делаешь dismiss, то повторно вызывается метод viewDidAppear у `SplashViewController`, а т.к токен еще не успели получить к этому моменту, то будет проскакивать экран авторизации. 

Причем еще будет проблема и с показом алерта, если произошла ошибка. Алерт не успеет показаться, т.к у тебя сразу будет вызван метод `viewDidAppear`, в котором сразу будет показан экран авторизации.

Как вариант решения этих проблем:
1. Добавить в UIBlockingProgressHUD свойство `isShowing`:
```
final class UIBlockingProgressHUD {
    
    static var isShowing: Bool = false
    
    private static var window: UIWindow? {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = windowScene.windows.first
        else {
            return nil
        }
        return window
    }

    static func show() {
        isShowing = true
        window?.isUserInteractionEnabled = false
        ProgressHUD.show()
    }

    static func dismiss() {
        isShowing = false
        window?.isUserInteractionEnabled = true
        ProgressHUD.dismiss()
    }
}
```

2. Во viewDidAppear добавить проверку:
```
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        guard UIBlockingProgressHUD.isShowing == false else { return }
        if let token = tokenStorage.token {
        ...
        }
    }
```

3. После того как будет показан алерт с ошибкой, на нажатие кнопки ok нужно добавить показ экрана авторизации

Commit:
processed optionality of createdAt property with ISO

Commit:
delete redundant self

Commit:
fix bugs
Splash flow
retain cycle in Images List
data format as in Figma
color of status bar
